docs/metadata.md

```diff
@@ -0,0 +1,582 @@
+
+The list of all [non-dev](#dependencydev) dependencies for this package. For interpreted
+languages, this is the list of runtime dependencies. For compiled languages, this is
+typically the list of everything needed for compilation. For Golang specifically, this
+is the list of all the [go-package][go-package-type] dependencies.
+
+Each dependency in the list is an object with a single attribute: `purl`.
+
+```json
+[
+  {"purl": "pkg:npm/axios@0.21.1"},
+  {"purl": "pkg:npm/follow-redirects@1.14.0"}
+]
+```
+
+Like the package purl, the purl of each dependency is intended to locate that dependency
+on the internet. Unlike packages, which are part of the git repository that Cachito
+processed, dependencies usually come from official package repositories such as PyPI or
+the NPM/Yarn registry. This allows Cachito to use more specialized purl types which
+convey more information about the name and version of the dependencies.
+
+Often, the JSON and purl representations of dependencies carry equivalent information,
+and it is possible to convert between them directly. Sometimes, type information can be
+lost, for example `npm` and `yarn` dependencies are both represented as `pkg:npm`. That
+is usually not important, in the NPM/Yarn case, the two are just different clients for
+the same package repository.
+
+Dependencies that do not come from official sources typically have to use more generic
+purl types, see [external dependencies](#external-dependencies).
+
+### sources
+
+For most package types, this is the list of all dependencies, i.e., a union of dev and
+non-dev dependencies. For Golang, this is the list of all the [gomod][go-package-type]
+dependencies.
+
+Much like the `dependencies` list, each dependency listed here is an object with
+a single `purl` attribute. The purls likewise follow the same logic as those in
+`dependencies`.
+
+### Special cases
+
+#### Golang packages
+
+Unlike most languages, Golang has no centralized package repository. In the Golang
+ecosystem, the name of a package must reflect its location on the internet. This allows
+Cachito to use the specialized `pkg:golang` purls even for packages, whereas for other
+languages Cachito can only afford to do so for dependencies.
+
+<details>
+<summary>Example</summary>
+
+```json
+{"name": "github.com/release-engineering/retrodep/v2", "version": "v2.1.1", "type": "go-package"}
+```
+```
+pkg:golang/github.com%2Frelease-engineering%2Fretrodep%2Fv2@v2.1.1
+```
+</details>
+
+One unfortunate downside of this approach is that if your repository is a fork intended
+for use with the [replace][gomod-replace] directive, the purl will be incorrect. Take
+for example the [openshift/kubernetes-apiserver][openshift-apiserver] repository. The
+name in `go.mod` is `k8s.io/apiserver`, which is the name of the package that this fork
+is meant to replace. That is expected as per the `replace` rules, but it also causes the
+purl to incorrectly reference the original rather than the fork.
```

Comment:
We may also have the opposite problem when somebody uses [replace](https://golang.org/ref/mod#go-mod-file-replace) for a dependency, e.g.
```
replace k8s.io/apiserver => github.com/openshift/kubernetes-apiserver
```
IIRC, Cachito takes the replacement name, so the purl ends up being `pkg:golang/github.com/openshift/kubernetes-apiserver`. Now, the purl semantics were never super clear to me, but one could make a good argument that such a purl is wrong, because `github.com/openshift/kubernetes-apiserver` is not a module that exists anywhere. The module is `k8s.io/apiserver`, the openshift fork is merely an alternative location for that module.

`go get` would seem to support that argument:
```
$ go get github.com/openshift/kubernetes-apiserver
go: github.com/openshift/kubernetes-apiserver upgrade => v0.0.0-20190820102804-33bf53f7c421
go get: github.com/openshift/kubernetes-apiserver@v0.0.0-20190820102804-33bf53f7c421: parsing go.mod:
	module declares its path as: k8s.io/apiserver
	        but was required as: github.com/openshift/kubernetes-apiserver
```
Then again, I don't know what a correct purl would be in this case. Probably a generic, non-golang one.